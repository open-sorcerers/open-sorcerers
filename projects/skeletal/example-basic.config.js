const R = require("ramda")
const { trace } = require("xtrace")

console.log("SKELETAL CONFIG")

module.exports = bones => {
  console.log("RAWR", bones)
  bones.pattern({
    name: "one",
    prompts: [{ type: "input", name: "name", message: "Name?" }],
    actions: [{ type: "add", template: "one.hbs" }]
  })

  bones.pattern({
    name: "two",
    prompts: [
      { type: "input", name: "description", message: "Description?" },
      { type: "confirm", name: "bool", message: "Vrai ou faux?" }
    ],
    actions: [{ type: "add", template: "two.hbs" }]
  })
  return { chimichanga: "yass" }
}
