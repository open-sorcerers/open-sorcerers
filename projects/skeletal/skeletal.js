'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var ramda = require('ramda');
var fluture = require('fluture');
var ensorcel = require('ensorcel');
require('handlebars');
var cosmiconfig = require('cosmiconfig');
var inquirer = require('inquirer');

var PLACEHOLDER = "üçõ";
var $ = PLACEHOLDER;
var bindInternal3 = function bindInternal3 (func, thisContext) {
  return function (a, b, c) {
    return func.call(thisContext, a, b, c);
  };
};
var some$1 = function fastSome (subject, fn, thisContext) {
  var length = subject.length,
      iterator = thisContext !== undefined ? bindInternal3(fn, thisContext) : fn,
      i;
  for (i = 0; i < length; i++) {
    if (iterator(subject[i], i, subject)) {
      return true;
    }
  }
  return false;
};

var curry = function (fn) {
  var test = function (x) { return x === PLACEHOLDER; };
  return function curried() {
    var arguments$1 = arguments;
    var argLength = arguments.length;
    var args = new Array(argLength);
    for (var i = 0; i < argLength; ++i) {
      args[i] = arguments$1[i];
    }
    var countNonPlaceholders = function (toCount) {
      var count = toCount.length;
      while (!test(toCount[count])) {
        count--;
      }
      return count
    };
    var length = (
      some$1(args, test) ?
        countNonPlaceholders(args) :
        args.length
    );
    function saucy() {
      var arguments$1 = arguments;
      var arg2Length = arguments.length;
      var args2 = new Array(arg2Length);
      for (var j = 0; j < arg2Length; ++j) {
        args2[j] = arguments$1[j];
      }
      return curried.apply(this, args.map(
        function (y) { return (
          test(y) && args2[0] ?
            args2.shift() :
            y
        ); }
      ).concat(args2))
    }
    return (
      length >= fn.length ?
        fn.apply(this, args) :
        saucy
    )
  }
};

var innerpipe = function (args) { return function (x) {
  var first = args[0];
  var rest = args.slice(1);
  var current = first(x);
  for (var a = 0; a < rest.length; a++) {
    current = rest[a](current);
  }
  return current
}; };
function pipe() {
  var arguments$1 = arguments;
  var argLength = arguments.length;
  var args = new Array(argLength);
  for (var i = 0; i < argLength; ++i) {
    args[i] = arguments$1[i];
  }
  return innerpipe(args)
}

var prop = curry(function (property, o) { return o && property && o[property]; });
var _keys = Object.keys;
var keys = _keys;

var propLength = prop("length");
var objectLength = pipe(keys, propLength);

var delegatee = curry(function (method, arg, x) { return (x[method](arg)); });

var filter = delegatee("filter");

function curryObjectN(arity, fn) {
  return function ŒªcurryObjectN(args) {
    var joined = function (z) { return ŒªcurryObjectN(Object.assign({}, args, z)); };
    return (
      args && Object.keys(args).length >= arity ?
        fn(args) :
        joined
    )
  }
}

var callWithScopeWhen = curry(function (effect, when, what, value) {
  var scope = what(value);
  if (when(scope)) { effect(scope); }
  return value;
});
var callBinaryWithScopeWhen = curry(function (effect, when, what, tag, value) {
  var scope = what(value);
  if (when(tag, scope)) { effect(tag, scope); }
  return value;
});

var always = function always() {
  return true;
};
var I = function I(x) {
  return x;
};

var callWhen = callWithScopeWhen($, $, I);
var call = callWithScopeWhen($, always, I);
var callWithScope = callWithScopeWhen($, always);
var callBinaryWhen = callBinaryWithScopeWhen($, $, I);
var callBinaryWithScope = callBinaryWithScopeWhen($, always);
var callBinary = callBinaryWithScopeWhen($, always, I);

var traceWithScopeWhen = callBinaryWithScopeWhen(console.log);
var traceWithScope = traceWithScopeWhen(always);
var inspect = traceWithScope;
var trace = inspect(I);
var traceWhen = callBinaryWithScopeWhen(console.log, $, I);

var segment = curryObjectN(3, function (_ref) {
  var _ref$what = _ref.what,
      what = _ref$what === void 0 ? I : _ref$what,
      _ref$when = _ref.when,
      when = _ref$when === void 0 ? always : _ref$when,
      tag = _ref.tag,
      value = _ref.value,
      effect = _ref.effect;
  if (when(tag, what(value))) {
    effect(tag, what(value));
  }
  return value;
});
var segmentTrace = segment({
  effect: console.log
});

var name = "skeletal";
var version = "0.0.0";
var description = "Build the bones of a project";
var main = "skeletal.js";
var module$1 = "skeletal.mjs";
var bin = {
	skeletal: "skeletal.js"
};
var repository = {
	type: "git",
	url: "https://github.com/open-sorcerers/skeletal.git"
};
var keywords = [
	"generator",
	"generate",
	"scaffolding",
	"scaffold",
	"template",
	"build",
	"pattern",
	"gen",
	"plop",
	"bones",
	"structure"
];
var author = "brekk";
var license = "MIT";
var devDependencies = {
	eslint: "^6.8.0",
	"eslint-config-sorcerers": "^0.0.2",
	jest: "^25.1.0",
	nps: "^5.9.12",
	prettier: "^1.19.1",
	"prettier-eslint": "^9.0.1",
	rollup: "^1.31.1"
};
var dependencies = {
	chalk: "^1.1.3",
	cosmiconfig: "^6.0.0",
	ensorcel: "^0.0.2",
	flexeca: "^0.0.1",
	fluture: "^12.2.0",
	handlebars: "^4.7.3",
	inquirer: "^7.0.4",
	ora: "^4.0.3",
	ramda: "^0.27.0",
	torpor: "^0.0.4",
	"yargs-parser": "^17.0.0"
};
var pkg = {
	name: name,
	version: version,
	description: description,
	main: main,
	module: module$1,
	bin: bin,
	repository: repository,
	keywords: keywords,
	author: author,
	license: license,
	devDependencies: devDependencies,
	dependencies: dependencies
};

var freeze = Object.freeze;
var own = function (z) { return Object.getOwnPropertyNames(z); };

function deepfreeze(o) {
  if (o === Object(o)) {
    if (!Object.isFrozen(o)) { freeze(o); }
    own(o).forEach(function (prop) {
      if (prop !== "constructor") { deepfreeze(o[prop]); }
    });
  }
  return o
}

var fork = ensorcel.tacit(2, fluture.fork);

var cosmicConfigurate = ramda.curry(function (ligament, cosmic) {
  var cancel = ramda.propOr(ramda.identity, "cancel", ligament);
  var futurize = ensorcel.futurizeWithCancel(cancel);
  var cosmicLoad = futurize(1, cosmic.load);
  var cosmicSearch = futurize(0, cosmic.search);
  return ramda.pipe(
    ramda.ifElse(ramda.pathOr(false, ["config", "configFile"]), cosmicLoad, function () { return cosmicSearch(); }
    ),
    ramda.map(
      ramda.pipe(
        ramda.propOr(ramda.identity, "config"),
        // ligasure ^^
        function (z) { return z(ligament); }
      )
    )
  )(ligament)
});

var UNSET = "%UNSET%";

var error = ramda.curry(function (ns, message, data) {
  var name = (pkg.name) + "@" + (pkg.version) + "-" + ns;
  var e = new Error(message);
  e.name = name;
  e.data = data;
  throw e
});

var getName = ramda.propOr(UNSET, "name");
var getPrompts = ramda.propOr(UNSET, "prompts");
var getActions = ramda.propOr(UNSET, "actions");
var ERROR = deepfreeze({
  EXPECTED_NAME_AND_MORE: error(
    "pattern__badinputs",
    "Expected name, prompts and actions properties to be given."
  )
});

var validatePatternAndSubmit = ramda.curry(function (bad, good, raw) { return ramda.pipe(
    ensorcel.box,
    ramda.ap([getName, getPrompts, getActions]),
    ramda.ifElse(
      ramda.any(ramda.equals(UNSET)),
      ramda.pipe(ERROR.EXPECTED_NAME_AND_MORE, bad),
      function (ref) {
        var name = ref[0];
        var prompts = ref[1];
        var actions = ref[2];

        return ({ name: name, prompts: prompts, actions: actions });
    }
    ),
    good
  )(raw); }
);

var pattern = ramda.curry(function (config, raw) {
  var cancel = ramda.propOr(ramda.identity, "cancel", config);
  var willPrompt = ensorcel.futurizeWithCancel(cancel, 1, inquirer.prompt);
  return [
    raw.name,
    ramda.pipe(
      ramda.chain(function (futurePattern) { return ramda.pipe(
          ramda.propOr([], "prompts"),
          ramda.map(willPrompt),
          ramda.reduce(
            function (left, right) { return ramda.chain(function (ll) { return ramda.map(ramda.mergeRight(ll), right); }, left); },
            fluture.resolve({})
          ),
          ramda.map(function (prompts) { return ramda.mergeRight(futurePattern, { prompts: prompts }); })
        )(futurePattern); }
      )
    )(
      new fluture.Future(function (bad, good) {
        validatePatternAndSubmit(bad, good, raw);
        return cancel
      })
    )
  ]
});

var pushInto = ramda.curry(function (into, fn) { return ramda.pipe(
    fn,
    call(function (x) { return into.push(x); })
  ); }
);

var saveKeyed = ramda.curry(function (struct, fn, input) {
  var ref = fn(input);
  var key = ref[0];
  var ff = ref[1];
  struct[key] = ff;
  return ff
});

var skeletal = function (config) {
  var parallelThreadMax = ramda.propOr(10, "threads", config);
  var isCancelled = false;
  /* const patterns = [] */
  var patterns = {};
  var cancel = function () {
    isCancelled = true;
  };
  // closured, for your safety
  var checkCancelled = function () { return isCancelled; };
  // wrap our composition steps with this so we can barf early
  var cancellable = ramda.unless(checkCancelled);
  // this is what the consumer sees as "bones" in the config file
  var ligament = {
    parallelThreadMax: parallelThreadMax,
    done: cancellable(function () { return ramda.pipe(
        ramda.toPairs,
        ramda.filter(function (ref) {
          var k = ref[0];

          return ramda.equals(ramda.propOr(false, "pattern", config), k);
        }),
        ramda.map(ramda.nth(1)),
        ramda.head
      )(patterns); }
    ),
    cancel: cancel,
    checkCancelled: checkCancelled,
    config: deepfreeze(config)
  };
  /* ligament.pattern = pushInto(patterns, pattern(ligament)) */
  ligament.pattern = saveKeyed(patterns, pattern(ligament));
  return ramda.pipe(
    ramda.propOr("skeletal", "namespace"),
    cosmiconfig.cosmiconfig,
    cancellable(cosmicConfigurate(ligament)),
    ramda.chain(
      ramda.cond([
        [checkCancelled, function () { return fluture.reject(new Error("CANCELLED")); }],
        [function () { return ramda.propOr(false, "pattern", config); }, ligament.done],
        [function () { return true; }, function () { return fluture.resolve({ patterns: ramda.keys(patterns) }); }]
      ])
    )
  )(config)
};

exports.cosmicConfigurate = cosmicConfigurate;
exports.deepfreeze = deepfreeze;
exports.fork = fork;
exports.pattern = pattern;
exports.pushInto = pushInto;
exports.saveKeyed = saveKeyed;
exports.skeletal = skeletal;
