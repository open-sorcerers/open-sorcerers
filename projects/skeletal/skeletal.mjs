import { curry as curry$1, propOr, identity, pipe as pipe$1, ifElse, pathOr, map, chain, ap, any, equals, reduce, mergeRight, unless } from 'ramda';
import { fork as fork$1, resolve, Future, parallel } from 'fluture';
import { tacit, futurizeWithCancel, box, j2 } from 'ensorcel';
import 'handlebars';
import { cosmiconfig } from 'cosmiconfig';
import { prompt } from 'inquirer';

var PLACEHOLDER = "üçõ";
var $ = PLACEHOLDER;
var bindInternal3 = function bindInternal3 (func, thisContext) {
  return function (a, b, c) {
    return func.call(thisContext, a, b, c);
  };
};
var some$1 = function fastSome (subject, fn, thisContext) {
  var length = subject.length,
      iterator = thisContext !== undefined ? bindInternal3(fn, thisContext) : fn,
      i;
  for (i = 0; i < length; i++) {
    if (iterator(subject[i], i, subject)) {
      return true;
    }
  }
  return false;
};

var curry = function (fn) {
  var test = function (x) { return x === PLACEHOLDER; };
  return function curried() {
    var arguments$1 = arguments;
    var argLength = arguments.length;
    var args = new Array(argLength);
    for (var i = 0; i < argLength; ++i) {
      args[i] = arguments$1[i];
    }
    var countNonPlaceholders = function (toCount) {
      var count = toCount.length;
      while (!test(toCount[count])) {
        count--;
      }
      return count
    };
    var length = (
      some$1(args, test) ?
        countNonPlaceholders(args) :
        args.length
    );
    function saucy() {
      var arguments$1 = arguments;
      var arg2Length = arguments.length;
      var args2 = new Array(arg2Length);
      for (var j = 0; j < arg2Length; ++j) {
        args2[j] = arguments$1[j];
      }
      return curried.apply(this, args.map(
        function (y) { return (
          test(y) && args2[0] ?
            args2.shift() :
            y
        ); }
      ).concat(args2))
    }
    return (
      length >= fn.length ?
        fn.apply(this, args) :
        saucy
    )
  }
};

var innerpipe = function (args) { return function (x) {
  var first = args[0];
  var rest = args.slice(1);
  var current = first(x);
  for (var a = 0; a < rest.length; a++) {
    current = rest[a](current);
  }
  return current
}; };
function pipe() {
  var arguments$1 = arguments;
  var argLength = arguments.length;
  var args = new Array(argLength);
  for (var i = 0; i < argLength; ++i) {
    args[i] = arguments$1[i];
  }
  return innerpipe(args)
}

var prop = curry(function (property, o) { return o && property && o[property]; });
var _keys = Object.keys;
var keys = _keys;

var propLength = prop("length");
var objectLength = pipe(keys, propLength);

var delegatee = curry(function (method, arg, x) { return (x[method](arg)); });

var filter = delegatee("filter");

function curryObjectN(arity, fn) {
  return function ŒªcurryObjectN(args) {
    var joined = function (z) { return ŒªcurryObjectN(Object.assign({}, args, z)); };
    return (
      args && Object.keys(args).length >= arity ?
        fn(args) :
        joined
    )
  }
}

var callWithScopeWhen = curry(function (effect, when, what, value) {
  var scope = what(value);
  if (when(scope)) { effect(scope); }
  return value;
});
var callBinaryWithScopeWhen = curry(function (effect, when, what, tag, value) {
  var scope = what(value);
  if (when(tag, scope)) { effect(tag, scope); }
  return value;
});

var always = function always() {
  return true;
};
var I = function I(x) {
  return x;
};

var callWhen = callWithScopeWhen($, $, I);
var call = callWithScopeWhen($, always, I);
var callWithScope = callWithScopeWhen($, always);
var callBinaryWhen = callBinaryWithScopeWhen($, $, I);
var callBinaryWithScope = callBinaryWithScopeWhen($, always);
var callBinary = callBinaryWithScopeWhen($, always, I);

var traceWithScopeWhen = callBinaryWithScopeWhen(console.log);
var traceWithScope = traceWithScopeWhen(always);
var inspect = traceWithScope;
var trace = inspect(I);
var traceWhen = callBinaryWithScopeWhen(console.log, $, I);

var segment = curryObjectN(3, function (_ref) {
  var _ref$what = _ref.what,
      what = _ref$what === void 0 ? I : _ref$what,
      _ref$when = _ref.when,
      when = _ref$when === void 0 ? always : _ref$when,
      tag = _ref.tag,
      value = _ref.value,
      effect = _ref.effect;
  if (when(tag, what(value))) {
    effect(tag, what(value));
  }
  return value;
});
var segmentTrace = segment({
  effect: console.log
});

var name = "skeletal";
var version = "0.0.0";
var description = "Build the bones of a project";
var main = "skeletal.js";
var module = "skeletal.mjs";
var bin = {
	skeletal: "skeletal.js"
};
var repository = {
	type: "git",
	url: "https://github.com/open-sorcerers/skeletal.git"
};
var keywords = [
	"generator",
	"generate",
	"scaffolding",
	"scaffold",
	"template",
	"build",
	"pattern",
	"gen",
	"plop",
	"bones",
	"structure"
];
var author = "brekk";
var license = "MIT";
var devDependencies = {
	eslint: "^6.8.0",
	"eslint-config-sorcerers": "^0.0.2",
	jest: "^25.1.0",
	nps: "^5.9.12",
	prettier: "^1.19.1",
	"prettier-eslint": "^9.0.1",
	rollup: "^1.31.1"
};
var dependencies = {
	chalk: "^1.1.3",
	cosmiconfig: "^6.0.0",
	ensorcel: "^0.0.2",
	flexeca: "^0.0.1",
	fluture: "^12.2.0",
	handlebars: "^4.7.3",
	inquirer: "^7.0.4",
	ora: "^4.0.3",
	ramda: "^0.27.0",
	torpor: "^0.0.4",
	"yargs-parser": "^17.0.0"
};
var pkg = {
	name: name,
	version: version,
	description: description,
	main: main,
	module: module,
	bin: bin,
	repository: repository,
	keywords: keywords,
	author: author,
	license: license,
	devDependencies: devDependencies,
	dependencies: dependencies
};

var freeze = Object.freeze;
var own = function (z) { return Object.getOwnPropertyNames(z); };

function deepfreeze(o) {
  if (o === Object(o)) {
    if (!Object.isFrozen(o)) { freeze(o); }
    own(o).forEach(function (prop) {
      if (prop !== "constructor") { deepfreeze(o[prop]); }
    });
  }
  return o
}

var fork = tacit(2, fork$1);

var cosmicConfigurate = curry$1(function (ligament, cosmic) {
  var cancel = propOr(identity, "cancel", ligament);
  var futurize = futurizeWithCancel(cancel);
  var cosmicLoad = futurize(1, cosmic.load);
  var cosmicSearch = futurize(0, cosmic.search);
  return pipe$1(
    ifElse(pathOr(false, ["config", "configFile"]), cosmicLoad, function () { return cosmicSearch(); }
    ),
    map(
      pipe$1(
        propOr(identity, "config"),
        // ligasure ^^
        function (z) { return z(ligament); }
      )
    ),
    chain(ligament.done)
  )(ligament)
});

var UNSET = "%UNSET%";

var error = curry$1(function (ns, message, data) {
  var name = (pkg.name) + "@" + (pkg.version) + "-" + ns;
  var e = new Error(message);
  e.name = name;
  e.data = data;
  throw e
});

var getName = propOr(UNSET, "name");
var getPrompts = propOr(UNSET, "prompts");
var getActions = propOr(UNSET, "actions");
var ERROR = deepfreeze({
  EXPECTED_NAME_AND_MORE: error(
    "pattern__badinputs",
    "Expected name, prompts and actions properties to be given."
  )
});

var validatePatternAndSubmit = curry$1(function (bad, good, raw) { return pipe$1(
    box,
    ap([getName, getPrompts, getActions]),
    ifElse(
      any(equals(UNSET)),
      pipe$1(ERROR.EXPECTED_NAME_AND_MORE, bad),
      function (ref) {
        var name = ref[0];
        var prompts = ref[1];
        var actions = ref[2];

        return ({ name: name, prompts: prompts, actions: actions });
    }
    ),
    good
  )(raw); }
);

var pattern = curry$1(function (config, raw) {
  var cancel = propOr(identity, "cancel", config);
  var willPrompt = futurizeWithCancel(cancel, 1, prompt);
  return pipe$1(
    chain(function (futurePattern) { return pipe$1(
        propOr([], "prompts"),
        map(willPrompt),
        reduce(function (left, right) {
          return chain(function (leftVal) {
            return map(function (rightVal) {
              return mergeRight(leftVal, rightVal)
            }, right)
          }, left)
        }, resolve({})),
        map(function (prompts) { return mergeRight(futurePattern, { prompts: prompts }); })
      )(futurePattern); }
    )
  )(
    new Future(function (bad, good) {
      validatePatternAndSubmit(bad, good, raw);
      return cancel
    })
  )
});

var pushInto = curry$1(function (into, fn) { return pipe$1(
    fn,
    call(function (x) { return into.push(x); })
  ); }
);

var skeletal = function (config) {
  var parallelThreadMax = propOr(10, "threads", config);
  var isCancelled = false;
  var patterns = [];
  var cancel = function () {
    isCancelled = true;
  };
  // closured, for your safety
  var checkCancelled = function () { return isCancelled; };
  // wrap our composition steps with this so we can barf early
  var cancellable = unless(checkCancelled);
  // this is what the consumer sees as "bones" in the config file
  var ligament = {
    parallelThreadMax: parallelThreadMax,
    done: cancellable(function () {
      var allPatterns = parallel(parallelThreadMax)(patterns);
      return allPatterns
    }),
    cancel: cancel,
    checkCancelled: checkCancelled,
    config: deepfreeze(config)
  };
  ligament.pattern = pushInto(patterns, pattern(ligament));
  return pipe$1(
    propOr("skeletal", "namespace"),
    cosmiconfig,
    cancellable(cosmicConfigurate(ligament)),
    map(call(function (x) { return console.log("what dis?", j2(x)); }))
  )(config)
};

export { cosmicConfigurate, deepfreeze, fork, pattern, pushInto, skeletal };
