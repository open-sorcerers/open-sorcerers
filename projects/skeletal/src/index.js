export { version, name } from "../package.json"
export * from "./skeletal"
export * from "./utils"
