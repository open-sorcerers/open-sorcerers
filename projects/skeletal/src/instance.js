import { bold } from "kleur"
import { pipe } from "ramda"
import PKG from "../package.json"
export const nameAndVersion = () => PKG.name + PKG.version
export const nameVersion = pipe(nameAndVersion, bold)
