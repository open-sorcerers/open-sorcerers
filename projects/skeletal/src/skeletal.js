import {
  __ as $,
  apply,
  curry,
  pipe,
  ifElse,
  identity as I,
  always as K,
  unless,
  map,
  filter,
  join,
  split,
  reduce,
  is,
  propOr,
  pathOr
} from "ramda"
import { trace } from "xtrace"
import { Future as F, fork as rawFork } from "fluture"
import { tacit, futurizeWithCancel } from "ensorcel"
import { compile } from "handlebars"
import { cosmiconfig } from "cosmiconfig"

export const fork = tacit(2, rawFork)

export const cosmicConfigurate = curry((ligament, cancel, conf) => {
  const futurize = futurizeWithCancel(cancel)
  const cosmicLoad = futurize(1, conf.load)
  const cosmicSearch = futurize(0, conf.search)

  return pipe(
    z => z || {},
    ifElse(propOr(false, "configFile"), cosmicLoad, () => cosmicSearch()),
    map(pipe(propOr(I, "config"), z => z(ligament)))
  )(conf)
})

export const skeletal = config => {
  let isCancelled = false
  const cancel = () => {
    isCancelled = true
  }
  const ligament = { cancel, isCancelled, pattern: config.pattern || I }
  const run = unless(() => isCancelled)
  return pipe(
    propOr("skeletal", "namespace"),
    cosmiconfig,
    run(cosmicConfigurate(ligament, cancel))
  )(config)
}
