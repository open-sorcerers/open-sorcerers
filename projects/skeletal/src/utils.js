import { fork as rawFork } from "fluture"
import {
  curry,
  replace,
  identity as I,
  pipe,
  split,
  map,
  when,
  includes,
  join
} from "ramda"
import { tacit } from "ensorcel"
const freeze = Object.freeze
const own = z => Object.getOwnPropertyNames(z)

export const deepfreeze = o => {
  if (o === Object(o)) {
    if (!Object.isFrozen(o)) freeze(o)
    own(o).forEach(prop => {
      if (prop !== "constructor") deepfreeze(o[prop])
    })
  }
  return o
}

const NM = "node_modules"

export const cutAfterString = curry((aa, bb) =>
  bb.slice(bb.indexOf(aa) + aa.length)
)

const unwrap = replace(")", "")

export const austereStack = when(
  I,
  pipe(
    split("\n"),
    map(
      when(
        includes(NM),
        /* z => "    at " + z.slice(z.indexOf(NM) + NM_LENGTH).replace(")", "") */
        z => "    at " + pipe(cutAfterString(NM), unwrap)(z)
      )
    ),
    join("\n")
  )
)
export const fork = tacit(2, rawFork)
