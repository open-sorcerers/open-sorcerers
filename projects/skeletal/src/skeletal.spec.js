import { skeletal } from "./skeletal"
import { fork } from "./utils"
import { nameVersion } from "./instance"

test("skeletal - basic", done => {
  fork(
    done,
    out => {
      expect(out).toMatchSnapshot()
      done()
    },
    skeletal({
      namespace: "example-basic"
    })
  )
})

test("skeletal - no config", done => {
  const namespace = "namespace" + Math.round(Math.random() * 1e5)
  fork(
    out => {
      expect(out).toEqual(
        `No config file found for namespace: "${namespace}". Try "bone --init ${namespace}"?`
      )
      done()
    },
    done,
    skeletal({ namespace })
  )
})
test("skeletal - initialize", done => {
  fork(
    done,
    eeee => {
      expect(eeee).toEqual(
        `ðŸ¦´ ${nameVersion()} - Wrote initial config file to my-cool-ns.config.js!\n\tRun 'bone -n my-cool-ns -p madlib' :)`
      )
      done()
    },
    skeletal({ init: "my-cool-ns" })
  )
})

test("skeletal - cancellation", done => {
  const cancel = fork(done, done, skeletal({ cancel: done }))
  cancel()
})
