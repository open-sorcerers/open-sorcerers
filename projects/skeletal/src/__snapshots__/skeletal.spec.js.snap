// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`skeletal - basic 1`] = `
"ðŸ¦´ [1mskeletal0.0.5[22m - Available patterns:
	- one
	- two
	- madlib"
`;
